packages/jest-haste-map/src/index.ts

```diff
@@ -339,30 +339,42 @@ class HasteMap extends EventEmitter {
 
   build(): Promise<InternalHasteMapObject> {
     if (!this._buildPromise) {
-      this._buildPromise = this._buildFileMap()
-        .then(data => this._buildHasteMap(data))
-        .then(hasteMap => {
+      this._buildPromise = (async () => {
+        const data = await this._buildFileMap();
+
+        // Persist when we don't know if files changed (changedFiles undefined)
+        // or when we know a file was changed or deleted.
+        let hasteMap: InternalHasteMap;
+        if (
+          data.changedFiles === undefined ||
+          data.changedFiles.size > 0 ||
+          data.removedFiles.size > 0
+        ) {
+          hasteMap = await this._buildHasteMap(data);
           this._persist(hasteMap);
+        } else {
+          hasteMap = data.hasteMap;
+        }
 
-          const rootDir = this._options.rootDir;
-          const hasteFS = new HasteFS({
-            files: hasteMap.files,
-            rootDir,
-          });
-          const moduleMap = new HasteModuleMap({
-            duplicates: hasteMap.duplicates,
-            map: hasteMap.map,
-            mocks: hasteMap.mocks,
-            rootDir,
-          });
-          const __hasteMapForTest =
-            (process.env.NODE_ENV === 'test' && hasteMap) || null;
-          return this._watch(hasteMap).then(() => ({
-            __hasteMapForTest,
-            hasteFS,
-            moduleMap,
-          }));
+        const rootDir = this._options.rootDir;
+        const hasteFS = new HasteFS({
+          files: hasteMap.files,
+          rootDir,
+        });
+        const moduleMap = new HasteModuleMap({
+          duplicates: hasteMap.duplicates,
+          map: hasteMap.map,
+          mocks: hasteMap.mocks,
+          rootDir,
         });
+        const __hasteMapForTest =
+          (process.env.NODE_ENV === 'test' && hasteMap) || null;
+        return this._watch(hasteMap).then(() => ({
```

Comment:
Just a quick observation, since `this_buildPromise` is an async function, is there a need to call `then` on `this._watch`? Shouldn't this be an await call followed by a return of the object?

Reply:
This didn't used to use async/await in the past, so this is probably why it looks this way. It can be changed to `await`.

Reply:
I'll change to `await`. Thanks for the feedback.